lib/user_interface.rb

```diff
@@ -91,28 +47,76 @@ def display_letter_header(letter)
   end
 
   def continue
-    output.print CONTINUE_MESSAGE
+    output.print Constants::CONTINUE_MESSAGE
     input.getch
   end
 
   def search_term
-    output.print SEARCH_MESSAGE
+    output.print Constants::SEARCH_MESSAGE
     collect_vaild_input { |user_input| validator.valid_string?(user_input) }
   end
 
   def search_again?
-    output.print ANOTHER_SEARCH_PROMPT
-    collect_vaild_input { |user_input| validator.valid_yes_no_answer?(user_input) }.downcase == YES_REPLY
+    boolen_choice?(Constants::ANOTHER_SEARCH_PROMPT)
+  end
+
+  def choose_contact(contacts)
+    display_all_with_index(contacts)
+    ask_for_index(contacts.length).to_i
+  end
+
+  def edit_field(contact)
+    display_contact(contact)
+    output.print Constants::FIELD_CHOICE_PROMPT
+
+    field = collect_field_key
+    output.print Constants::FIELDS_TO_PROMPTS[field]
+    value = collect_valid_field(field)
+    { field => value }
+  end
+
+  def update_another_field?
+    boolen_choice?(Constants::ANOTHER_EDIT_PROMPT)
+  end
+
+  def update_another_contact?
+    boolen_choice?(Constants::ANOTHER_UPDATE_PROMPT)
   end
 
   private
 
+  def ask_for_index(array_length)
+    output.print Constants::CONTACT_INDEX_PROMPT
+    collect_vaild_input { |user_input| validator.valid_index?(user_input, array_length) }
+  end
+
+  def display_all_with_index(contacts)
+    contacts.each_with_index do |contact, index|
+      output.print "[#{index}]"
+      display_contact(contact)
+    end
+  end
+
+  def collect_field_key
+    field_name = collect_vaild_input { |user_input| validator.valid_field_name?(user_input) }
+    Constants::FIELDS_TO_DISPLAY_NAMES.key(field_name.capitalize + ': ')
```

Comment:
:thinking: why do you do this? The objective of the input collector is to return a value in the format that you want it. 

In other words, `collect_valid_input` should already return a valid field name as a string. Then you can return it symbolised from here. This should suffice:


```ruby
  def collect_field_key
    collect_valid_input { |user_input| validator.valid_field_name?(user_input) }.downcase.to_sym
  end
```

Use symbols, not strings, in code.

* what if you change the structure of the display name and it is not anymore the field capitalized with a colon?
* what if you change the language and you don't have **phone/Phone** anymore, but rather **phone/Número de teléfono**? This will fail.

Reply:
I've had to put a disgusting level of behaviour in the user interface to do this. I'm concerned the new code I've written to put the match in should live outside the UI.

Reply:
The new code is OK :ok_hand: 
it makes it more visible what you are trying to do, the previous version was more confusing.

Just a small refactoring suggestion:
Instead of using `join` on `.values`, use `first`.

You will always just get one result, but if for whatever strange coincidence you get two matches, `join` will have unexpected results, while `first` will always give you the first result.

With that said, here is a simplified version:
```ruby
  def convert_name_to_key(field_name)
    Constants::FIELDS_TO_DISPLAY_NAMES.select { |_, name| name.match?(/#{field_name}/i) }.keys.first
  end
```